trillian/migrillian/main.go

```diff
@@ -0,0 +1,171 @@
+
+	startIndex = flag.Int64("start_index", 0, "CT log index to start scanning at")
+	endIndex   = flag.Int64("end_index", 0, "CT log index to end scanning at (non-inclusive, 0 = end of log)")
+
+	quiet = flag.Bool("quiet", false, "Don't print out extra logging messages")
+)
+
+// trillianTreeClient is a means of communicating with a Trillian log tree.
+type trillianTreeClient struct {
+	client trillian.TrillianLogClient
+	logID  int64
+}
+
+// addSequencedLeaves converts a batch of CT log entries into Trillian log
+// leaves and submits them to Trillian via AddSequencedLeaves API.
+func (c *trillianTreeClient) addSequencedLeaves(ctx context.Context, b *scanner.EntryBatch) error {
+	leaves := make([]*trillian.LogLeaf, len(b.Entries))
+	for i, e := range b.Entries {
+		logEntry, err := toLogEntry(b.Start+int64(i), &e)
+		if err != nil {
+			return fmt.Errorf("failed to build LogEntry: %v", err)
+		}
+		leaf, err := buildLogLeaf(logEntry)
+		if err != nil {
+			return fmt.Errorf("failed to build LogLeaf: %v", err)
+		}
+		leaves[i] = leaf
+	}
+
+	req := trillian.AddSequencedLeavesRequest{LogId: c.logID, Leaves: leaves}
+	rsp, err := c.client.AddSequencedLeaves(ctx, &req)
+	if err != nil {
+		return fmt.Errorf("AddSequencedLeaves(): %v", err)
+	} else if rsp == nil {
+		return errors.New("missing AddSequencedLeaves response")
+	}
+	// TODO(pavelkalinnikov): Check rsp.Results statuses.
+	return nil
+}
+
+// logEntrySubmitter is a worker function which takes CT log entry batches from
+// the channel and processes them. Terminates when the channel is closed.
+func logEntrySubmitter(ctx context.Context, c trillianTreeClient, batches <-chan scanner.EntryBatch) {
+	for b := range batches {
+		// TODO(pavelkalinnikov): Retry with backoff on errors.
+		err := c.addSequencedLeaves(ctx, &b)
+		if *quiet {
+			continue
+		}
+		end := b.Start + int64(len(b.Entries))
+		if err != nil {
+			log.Printf("Failed to add batch [%d, %d): %v\n", b.Start, end, err)
+		} else {
+			log.Printf("Added batch [%d, %d)\n", b.Start, end)
+		}
+	}
+}
+
+func main() {
+	flag.Parse()
+
+	transport := &http.Transport{
+		TLSHandshakeTimeout:   30 * time.Second,
+		ResponseHeaderTimeout: 30 * time.Second,
+		MaxIdleConnsPerHost:   10,
+		DisableKeepAlives:     false,
+		MaxIdleConns:          100,
+		IdleConnTimeout:       90 * time.Second,
+		ExpectContinueTimeout: 1 * time.Second,
+	}
+
+	ctClient, err := client.New(*ctLogURI, &http.Client{
+		Timeout:   10 * time.Second,
+		Transport: transport,
+	}, jsonclient.Options{})
+	if err != nil {
+		log.Fatalf("Failed to create client for source log: %v", err)
+	}
+
+	opts := &scanner.FetcherOptions{
+		BatchSize:     *batchSize,
+		ParallelFetch: *parallelFetch,
+		StartIndex:    *startIndex,
+		EndIndex:      *endIndex,
+		Quiet:         *quiet,
+	}
+	fetcher := scanner.NewFetcher(ctClient, opts)
+
+	bufferSize := 10 * *parallelSubmit
+	batches := make(chan scanner.EntryBatch, bufferSize)
+
+	fmt.Println("Dialing Trillian...")
```

Comment:
Should these be `glog`?

Reply:
Possibly glog.Info but some of our binaries do use stdout. Is someone likely to be watching this command run on a terminal?